Add ability to preview decoded function call data

ifavo opened the pull request:
When signing a transaction the data is hex encoded and not understandable from a human perspective. The situation allows to pretend to users something different than what will be signed.

B32 has a registry of many relevant ABI definitions. Using B32 and the clause data the transaction details can be decoded in a more human friendly format.

This pull request is a modification that will:

1. extract function signature from call data
2. lookup the ABI definition from b32
3. if found, try to decode the clause data using the ABI
4. display a decoded version

It is possible that the function name or parameter names are not identical due the same signature existing in multiple registry entries. It will still provide a better experience and help prevent signing of malicious transactions.

The decoded data is shown at the clause inspection below data:

![image](https://user-images.githubusercontent.com/407503/203303227-a0d3969a-ffb2-4846-b968-669aa04d3bb9.png)

XJWX89 commented:
hi  @ifavo ,Thank you for your submission, I have changed the style,  layout and  added utf8 decoding.

below is the sketch of clause inspection page view
![SCR-20221129-g8u](https://user-images.githubusercontent.com/12961554/204433317-d3daa65a-b7bd-4f64-bc80-2de868055d50.png)

Feel free to give your comments

ifavo commented:
I like that it saves screen space. It looks more like the explorer handling. I prefer your suggestion.
Is there a tab component already included that can be used?

AsbertMa commented:
>Yes,[ the UI lab](https://v1.quasar.dev/vue-components/tabs). Add the component to the _components_ array in **quasar.conf.js**

ifavo commented:
I was more wondering about the use in the app, I think I've not seen it applied in the app yeet.
You did not mention if you want to do the change yourself so I just modified it to use tabs.

The tabs by default don't seem to have the ability to be aligned left, so I left it centered. I am not familiar with vue and quasar, so maybe I am not seeing it. I personally would love to see it left aligned because it would fit better with the above list:

![Bildschirmaufnahme 2022-11-30 um 06 31 52](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif)

libotony commented:
Hi @ifavo, `connex` allows developer supply ABI, https://docs.vechain.org/connex/api.html#acquire-a-signing-service. You can get it from `TxMessage.abi`. So I would suggest using dev supplied ABI if it presents, otherwise fetch from b32. Here is an example you can try, https://codepen.io/libotony/pen/zYoPeaY

ifavo commented:
@libotony Thank you! I did not know about that.

I modified it to use it first and use b32 on error.

AsbertMa commented:
> > 是的，[UI 实验室](https://v1.quasar.dev/vue-components/tabs)。将组件添加到 quasar.conf.js 中的 components 数组
> 
> 我更想知道在应用程序中的使用，我想我还没有看到它在应用程序中应用。
> 
> > 我更改了样式、布局并添加了 utf8 解码。
> > 下面是条款检查页面视图的草图
> 
> 您没有提到是否要自己进行更改，所以我只是将其修改为使用制表符。
> 
> 默认情况下，选项卡似乎无法左对齐，所以我将其居中。我对 vue 和 quasar 不熟悉，所以可能我没有看到它。我个人很乐意看到它左对齐，因为它更适合上面的列表：
> 
> ![Bildschirmaufnahme 2022-11-30 嗯 06 31 52](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif) [ ![Bildschirmaufnahme 2022-11-30 嗯 06 31 52](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif) ](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif) [ ](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif)

AsbertMa commented:
> > Yes,[ the UI lab](https://v1.quasar.dev/vue-components/tabs). Add the component to the components array in quasar.conf.js
> 
> I was more wondering about the use in the app, I think I've not seen it applied in the app yet.
> 
> > I have changed the style, layout and added utf8 decoding.
> > below is the sketch of clause inspection page view
> 
> You did not mention if you want to do the change yourself so I just modified it to use tabs.
> 
> The tabs by default don't seem to have the ability to be aligned left, so I left it centered. I am not familiar with vue and quasar, so maybe I am not seeing it. I personally would love to see it left aligned because it would fit better with the above list:
> 
> ![Bildschirmaufnahme 2022-11-30 um 06 31 52](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif) [ ![Bildschirmaufnahme 2022-11-30 um 06 31 52](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif) ](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif) [ ](https://user-images.githubusercontent.com/407503/204715799-d1bc0d43-4130-496c-802c-f8d901a575d9.gif)

I did a pull request for the tab's style ( align left & color ), the way inject the components. Please review it

libotony commented:
> @libotony Thank you! I did not know about that.
> 
> I modified it to use `.abi` first and use b32 if decoding fails.

I would suggest using the developer-supplied ABI specifically if present, do not load from b32. What do you think?

ifavo commented:
> > @libotony Thank you! I did not know about that.
> > I modified it to use `.abi` first and use b32 if decoding fails.
> 
> I would suggest using the developer-supplied ABI specifically if present, do not load from b32. What do you think?

It is doing this as far as I tested it. Using your example I've modified the clauses ABIs input names and verified that the ABIs names are shown in the decoded tab.

Here is the specific `if`:
* https://github.com/vechain/sync2/pull/279/commits/a671bff633d28a01c17ac735dec1e7e97767c3e7#diff-c9d4a60aac4efba79ccbd3c4a16385557ec68c14fde71d38b91da557fb7c0f5cR131-R137

If it fails (an invalid ABI given) it will continue to lookup on b32.

ifavo commented:
> I did a pull request for the tab's style ( align left & color ), the way inject the components. Please review it

@AsbertMa I merged it, thanks a lot! It improves the styling but alignment still does not work, it seems it was implemented in the used version.

libotony commented:
> > > @libotony Thank you! I did not know about that.
> > > I modified it to use `.abi` first and use b32 if decoding fails.
> > 
> > 
> > I would suggest using the developer-supplied ABI specifically if present, do not load from b32. What do you think?
> 
> It is doing this as far as I tested it. Using your example I've modified the clauses ABIs input names and verified that the ABIs names are shown in the decoded tab.
> 
> Here is the specific `if`:
> 
> * [a671bff#diff-c9d4a60aac4efba79ccbd3c4a16385557ec68c14fde71d38b91da557fb7c0f5cR131-R137](https://github.com/vechain/sync2/commit/a671bff633d28a01c17ac735dec1e7e97767c3e7#diff-c9d4a60aac4efba79ccbd3c4a16385557ec68c14fde71d38b91da557fb7c0f5cR131-R137)
> 
> If it fails (an invalid ABI given) it will continue to lookup on b32.

I would prefer to stop looking on b32 if decoding failed(invalid ABI given).

ifavo commented:
> > It is doing this as far as I tested it. Using your example I've modified the clauses ABIs input names and verified that the ABIs names are shown in the decoded tab.
> > Here is the specific `if`:
> > 
> > * [a671bff#diff-c9d4a60aac4efba79ccbd3c4a16385557ec68c14fde71d38b91da557fb7c0f5cR131-R137](https://github.com/vechain/sync2/commit/a671bff633d28a01c17ac735dec1e7e97767c3e7#diff-c9d4a60aac4efba79ccbd3c4a16385557ec68c14fde71d38b91da557fb7c0f5cR131-R137)
> > 
> > If it fails (an invalid ABI given) it will continue to lookup on b32.
> 
> I would prefer to stop looking on b32 if decoding failed(invalid ABI given).

I would agree in a good world scenario.  
My main reason for this PR was to provide users the ability to understand whats happening to stop potential malicious transactions:

- When a bad actor initiates a transaction with an invalid ABI
- Then the decoded version will not show
- And the user can not see a potential decoded version of the data

What do you think?

libotony commented:
> > > It is doing this as far as I tested it. Using your example I've modified the clauses ABIs input names and verified that the ABIs names are shown in the decoded tab.
> > > Here is the specific `if`:
> > > 
> > > * [a671bff#diff-c9d4a60aac4efba79ccbd3c4a16385557ec68c14fde71d38b91da557fb7c0f5cR131-R137](https://github.com/vechain/sync2/commit/a671bff633d28a01c17ac735dec1e7e97767c3e7#diff-c9d4a60aac4efba79ccbd3c4a16385557ec68c14fde71d38b91da557fb7c0f5cR131-R137)
> > > 
> > > If it fails (an invalid ABI given) it will continue to lookup on b32.
> > 
> > 
> > I would prefer to stop looking on b32 if decoding failed(invalid ABI given).
> 
> I would agree in a good world scenario. My main reason for this PR was to provide users the ability to understand whats happening to stop potential malicious transactions:
> 
> * When a bad actor initiates a transaction with an invalid ABI
> * Then the decoded version will not show
> * And the user can not see a potential decoded version of the data
> 
> What do you think?

I see, you're right.  Look up on b32 reveals more potential.
